velox/functions/lib/Re2Functions.cpp

```diff
@@ -388,8 +388,143 @@ bool matchExactPattern(
+}
+
+bool matchRelaxedFixedAscii(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kLiteralString &&
+        std::memcmp(
+            input.data() + start + subPattern.start,
+            patternMetadata.fixedPattern().data() + subPattern.start,
+            subPattern.length) != 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool matchRelaxedFixedUnicode(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  auto cursor = start;
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kSingleCharWildcard) {
+      // Match every single char wildcard.
+      for (auto i = 0; i < subPattern.length; i++) {
+        if (cursor >= input.size()) {
+          return false;
+        }
+
+        auto chrOffset = utf8proc_char_length(input.data() + cursor);
+        // Skip bad byte if we get utf length < 0.
+        auto numBytes = UNLIKELY(chrOffset < 0) ? 1 : chrOffset;
+        cursor += numBytes;
+      }
+    } else {
+      const auto currentLength = subPattern.length;
+      if (cursor + currentLength > input.size() ||
+          std::memcmp(
+              input.data() + cursor,
+              patternMetadata.fixedPattern().data() + subPattern.start,
+              currentLength) != 0) {
+        return false;
+      }
+
+      cursor += currentLength;
+    }
+  }
+
+  return true;
+}
+
+// Match the input(from the position of start) with relaxed pattern forward.
+template <bool isAscii>
+bool matchRelaxedFixedPatternForward(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  if constexpr (isAscii) {
+    return matchRelaxedFixedAscii(input, patternMetadata, start);
+  } else {
+    return matchRelaxedFixedUnicode(input, patternMetadata, start);
+  }
+}
+
+// Match the input(from the position of start) with relaxed pattern backward.
+// Unlike matchRelaxedFixedPatternForward which has different path for utf8 and
+// ascii, this function only have implementation for utf8 because only utf8
+// input use this function.
+bool matchRelaxedFixedBackward(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if ((start + 1) < patternMetadata.length()) {
+    return false;
+  }
+
+  const auto& subPatterns = patternMetadata.subPatterns();
+  auto cursor = start;
+  for (int32_t i = subPatterns.size() - 1; i >= 0; i--) {
+    if (cursor < 0) {
+      return false;
+    }
+
+    const auto& subPattern = subPatterns[i];
+    const auto& kind = subPattern.kind;
```

Comment:
store enums and primitive types by value

kind variable is used only once; consider removing

currentStart is only used in else block; consider moving it there

currentLength is used in both if and else blocks, but usage in if block is redundant; the code will be easier to read as 

auto charsToSkip = subPattern.length